### ISSUE TITLE
Dbus-mpris: Couldn't fetch metdata from spotify

### ISSUE THREAD

**Issue by Glamhoth** [Spotifyd/spotifyd#121]
Spotifyd compiled with mpris support has a problem with fetching metadata. When listening to a song it logs this:

> Couldn't fetch metadata from spotify: Err(Error(Msg("convert result failed, content \"{
>
>\*a lot of metadata stuff, but seems to be correct\*
>
>State { next_error: Some(ErrorImpl { code: Message("invalid value: integer `-6973076`, expected u32"), line: 3, column: 26 }), backtrace: None }))

That value decreases each second.

Because of this problem mpris-ctl doesn't display any info other that 'unknow - unknown - unknown' (so at least it sees that *something* is playing).

System info:
Arch Linux 4.15.14-1-ARCH x86_64
rustc 1.25.0

edit:
For some reason it doesn't happen everytime, but mprist always sees it as 'unknown'. Althrough play/pause/next/etc. work fine.

**Comment by mikesmiffy128** [comment]
I have a somewhat similar issue at the moment, but it happens with playerctl, regardless of which command I use. Here's the output from `playerctl -p spotifyd status`:

```
12:50:50 [INFO] Couldn't fetch metadata from spotify: Err(Error(Msg("send request failed, http code:400 Bad Request, error message:<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 400 (Bad Request)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>400.</b> <ins>That’s an error.</ins>\n  <p>Your client has issued a malformed or illegal request.  <ins>That’s all we know.</ins>\n"), State { next_error: None, backtrace: None }))
parameters: Object({})

response: Response { url: "https://api.spotify.com/v1/me/player/currently-playing", status: BadRequest, headers: {"Content-Type": "text/html; charset=UTF-8", "Referrer-Policy": "no-referrer", "Content-Length": "1555", "Date": "Tue, 08 May 2018 11:50:50 GMT"} }
content: "<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 400 (Bad Request)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>400.</b> <ins>That’s an error.</ins>\n  <p>Your client has issued a malformed or illegal request.  <ins>That’s all we know.</ins>\n"
parameters: Object({})

response: Response { url: "https://api.spotify.com/v1/me/player", status: BadRequest, headers: {"Content-Type": "text/html; charset=UTF-8", "Referrer-Policy": "no-referrer", "Content-Length": "1555", "Date": "Tue, 08 May 2018 11:50:51 GMT"} }
content: "<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 400 (Bad Request)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>400.</b> <ins>That’s an error.</ins>\n  <p>Your client has issued a malformed or illegal request.  <ins>That’s all we know.</ins>\n"
```

My best guess as to what's going on is that there's some kind of breakage in `rspotify`. I haven't yet tracked down what exactly is happening to cause a bad request, and I don't really know Rust so I'm probably not the best person to figure that out. Maybe there's something up with the authentication but that's just a random guess.

Just compiled spotifyd-git from AUR today so I should be running from HEAD; I compiled with rustc 1.25 and cargo 0.26.

**Comment by mariusor** [comment]
@SimonPersson to check the raw output from DBus without using intermediaries like mpris-ctl or playerctl you can use the dbus-send cli tool:
```
dbus-send --print-reply --type=method_call  --dest=org.mpris.MediaPlayer2.spotifyd /org/mpris/MediaPlayer2 org.freedesktop.DBus.Properties.GetAll string:org.mpris.MediaPlayer2.Player
```

Also you can check the emitted signals using `dbus-monitor`

**Comment by SirWindfield** [comment]
This is probably fixed in v0.2.18

**Comment by mikesmiffy128** [comment]
Yep, only just got round to testing it, but appears to be working perfectly, thanks. Only small quirk is that it doesn't seem to respond until it's been started/stopped at least once already?? I'd need to mess with it more to be certain...
